packages/grid/x-data-grid-pro/src/DataGridPro/useDataGridProProps.ts

```diff
@@ -31,39 +32,43 @@ export const DATA_GRID_PRO_PROPS_DEFAULT_VALUES: DataGridProPropsWithDefaultValu
 };
 
 export const useDataGridProProps = <R extends GridValidRowModel>(inProps: DataGridProProps<R>) => {
-  const themedProps = useThemeProps({ props: inProps, name: 'MuiDataGrid' });
+  const { components, componentsProps, ...themedProps } = useThemeProps({
+    props: inProps,
+    name: 'MuiDataGrid',
+  });
 
   const localeText = React.useMemo(
     () => ({ ...GRID_DEFAULT_LOCALE_TEXT, ...themedProps.localeText }),
     [themedProps.localeText],
   );
 
-  const components = React.useMemo<GridProSlotsComponent>(() => {
-    const overrides = themedProps.components;
+  const slots = React.useMemo<UncapitalizedGridProSlotsComponent>(() => {
+    const uncapitalizedDefaultSlots = uncapitalizeObjectKeys(
+      DATA_GRID_PRO_DEFAULT_SLOTS_COMPONENTS,
+    )!;
+    const overrides = themedProps.slots ?? (components ? uncapitalizeObjectKeys(components) : null);
 
     if (!overrides) {
-      return { ...DATA_GRID_PRO_DEFAULT_SLOTS_COMPONENTS };
+      return { ...uncapitalizedDefaultSlots };
     }
 
-    const mergedComponents = {} as GridProSlotsComponent;
-
-    type GridSlots = keyof GridProSlotsComponent;
-    Object.entries(DATA_GRID_PRO_DEFAULT_SLOTS_COMPONENTS).forEach(([key, defaultComponent]) => {
-      mergedComponents[key as GridSlots] =
-        overrides[key as GridSlots] === undefined ? defaultComponent : overrides[key as GridSlots];
-    });
-
-    return mergedComponents;
-  }, [themedProps.components]);
+    type GridSlot = keyof UncapitalizedGridProSlotsComponent;
+    return Object.entries(uncapitalizedDefaultSlots).reduce((acc, [key, defaultComponent]) => {
+      const override = overrides[key as GridSlot];
+      const component = override !== undefined ? override : defaultComponent;
+      return { ...acc, [key as GridSlot]: component };
+    }, {} as UncapitalizedGridProSlotsComponent);
+  }, [components, themedProps.slots]);
 
   return React.useMemo<DataGridProProcessedProps<R>>(
     () => ({
       ...DATA_GRID_PRO_PROPS_DEFAULT_VALUES,
       ...themedProps,
       localeText,
-      components,
+      slots,
+      slotProps: themedProps.slotProps ?? componentsProps,
```

Comment:
That's a really good suggestion 🎉 , so either of `components/componentsProps` and `slots/slotProps` would work and not both.

Thinking more on this, it could have one minor issue too. In theory, the above-mentioned props should be passable (and users will be free to pass any combo of props from `components/slotProps`, `slots/componentsProps`, and `slots/slotProps`).

If we restrict cross-props usage or bind `components <==> componentsProps` and `slots <==> slotProps`, the cross-usage scenarios (like the one you mentioned above) will not work without users getting many clues why (other than the deprecation warning for one of the props shown in IDE).

Should we generate a warning to warn users about this mismatch to cater this?

> it would work in v6 but will break in v7.

Exactly, how about we allow cross-usage scenarios now and treat them in v7 under `breaking changes` category so the users will anyways have to remove them (the same for `components/components` or `slots/componentsProps`)?

Reply:
I'll add my 5 cents: IMHO, it doesn't seem like a big deal that we'd allow such scenario. Honestly, to me, the headache making sure that it wouldn't happen seems bigger than the potential benefit we'd get. 🤔 
Providing a codemod taking care of migration from `components` to `slots` and deprecating `components` does seem sufficient. The rest is up to individual users and their needs to decide how and when to migrate.

Reply:
The main use case I had in mind is:
- you're using `components` (which is deprecated)
- you copy a snippet with `slotProps` from the new docs
- you end up using both APIs

But I'm fine with ignoring it as it's an edge case :+1: